lib/core/init.js

```diff
@@ -1,57 +1,229 @@
 
-		// console.log('get options end: ' + Date());
-		for (var thisModuleID in modules) {
-			if (typeof modules[thisModuleID] === 'object') {
-				if (typeof modules[thisModuleID].beforeLoad === 'function') {
-					modules[thisModuleID].beforeLoad();
-				}
-			}
-		}
-		RESUtils.addBodyClasses();
 
-		// apply style...
-		RESUtils.addStyle(RESUtils.css);
-		// clear out css cache...
-		RESUtils.css = '';
+	return collection[type][id];
+};
+
+RESUtils.init._watch = function() {
+	RESUtils.init._watch.storage();
+	RESUtils.init._watch.documentElementReady();
+	RESUtils.init._watch.headReady();
+	$(RESUtils.init._watch.bodyReady);
+	window.addEventListener('load', RESUtils.init._watch.contentLoaded, false);
+};
+
+RESUtils.init._watch.storage = function() {
+	var def = RESStorage.setup();
+	RESUtils.init.await._deferred('storage', '_all', def);
+}
+
+RESUtils.init._watch.documentElementReady = function() {
+	if (document && document.documentElement && document.documentElement.classList) {
+		document.html = document.documentElement;
+		RESUtils.init.await._deferred('documentElementReady').resolve();
+	} else {
+		setTimeout(RESUtils.init._watch.documentElementReady, 1);
+	}
+};
+
+
+RESUtils.init._watch.headReady = function() {
+	if (document && document.head) {
+		RESUtils.init.await._deferred('documentElementReady').resolve();
+		RESUtils.init.await._deferred('documentHeadReady').resolve();
+	} else {
+		setTimeout(RESUtils.init._watch.headReady, 1);
+	}
+};
+
+RESUtils.init._watch.bodyReady = function() {
+	if (document && document.body && document.body.classList) {
+		RESUtils.init.await._deferred('documentElementReady').resolve();
+		RESUtils.init.await._deferred('documentHeadReady').resolve();
+		RESUtils.init.await._deferred('documentBodyReady').resolve();
 	} else {
-		setTimeout(RESdoBeforeLoad, 1);
+		setTimeout(RESUtils.init._watch.bodyReady, 1);
+	}
+};
+
+RESUtils.init._watch.contentLoaded = function() {
+	RESUtils.init.await._deferred('documentContentLoaded').resolve();
+};
+
+RESUtils.init._watch.initReadyCheck = function() {
+	// TODO: tidy this up, use await.documentHeadReady
+	if (
+		(!RESStorage.isReady) ||
+		(typeof document.body === 'undefined') ||
+		(!document.html) ||
+		(typeof document.html.classList === 'undefined')
+	) {
+		setTimeout(RESInitReadyCheck, 50);
+	} else {
+		RESUtils.runtime.RESInitReadyCheck(RESInit);
 	}
 }
 
-function RESInit() {
-	// body should be present now, add those classes.
-	RESUtils.addBodyClasses();
 
-	// if RESStorage isn't fully loaded, and _beforeLoadComplete isn't true,
-	// then wait. It means we haven't read all of the modules' options yet.
-	if (!RESStorage.isReady || !_beforeLoadComplete) {
-		setTimeout(RESInit, 10);
+RESUtils.init._loadModules = function () {
+	if (!RESUtils.regexes.all.test(location.href)) {
+		// added this if statement because some people's Greasemonkey "include" lines are getting borked or ignored, so they're calling RES on non-reddit pages.
```

Comment:
There's also a "migrate settings from Greasemonkey" bit still in browsersupport-firefox.js but dunno if it actually works..